matching-service/controllers/MatchingController.js

```diff
@@ -1,322 +1,110 @@
-    try {
-        const numInterviews = await Interview.countDocuments({});
-        var res = {
-            status: responseStatus.SUCCESS,
-            data: {
-                count: numInterviews
-            }
-        };
-        return res;
-    } catch (err) {
-        var res = {
-            status: responseStatus.ERROR,
-            message: mongoErrMsgs.readError(err)
-        };
-        return res;
-    }
-}
-
-// Get interview of user (if it exist)
-async function getInterview(email) {
-    try {
-        const interview = await Interview.findOne({
-            $or: [
-                { email1: email },
-                { email2: email }
-            ]
-        });
-
-        if (!interview) {
-            var res = {
-                status: responseStatus.NOT_FOUND,
-                message: clientErrMsgs.NO_INTERVIEW_FOUND_ERR
-            }
-            return res;
-        }
-
-        var partnerEmail = interview.email1;
-        if (partnerEmail == email) {
-            partnerEmail = interview.email2;
-        }
-
-        var secondsPassed = Math.floor((new Date().getTime() - interview.timeCreated.getTime()) / 1000);
-        const secondsLeft = 3600 - secondsPassed;
-
-        var res = {
-            status: responseStatus.SUCCESS,
-            data: {
-                partnerEmail: partnerEmail,
-                interviewId: interview.interviewId,
-                secondsLeft: secondsLeft
-            }
-        };
-
-        return res;
-    } catch (err) {
-        var res = {
-            status: responseStatus.ERROR,
-            message: mongoErrMsgs.readError(err)
-        };
-        return res;
-    }
+async function searchMatch(socket, io, email, difficulty, jwtToken, userId) {
+  var authRes = authJwt.verifyToken(jwtToken, userId, socket);
+
+  // Not authorised
+  if (authRes.status === responseStatus.UNAUTHORIZED) {
+    socket.emit("unauthorized", authRes);
+    console.log("Unauthorized");
+    return;
+  }
+
+  // Authorised
+  if (!requestHelpers.isValidDifficulty(difficulty)) {
+    var res = {
+      status: responseStatus.BAD_REQUEST,
+      message: clientErrMsgs.INVALID_DIFFICULTY_ERR,
+    };
+    socket.emit("badRequest", res);
+    console.log(message);
+    return;
+  }
+
+  // Check if this user has already request pending
+  // Mongo Timing of the Delete Operation is every 60 secs
+  const dupRequest = await Match.findOne({ email: email });
+  if (dupRequest && Date.now - dupRequest.timeCreated <= 30) {
+    var res = {
+      status: responseStatus.BAD_REQUEST,
+      message: clientErrMsgs.DUPLICATED_REQUEST_ERR,
+    };
+    socket.emit("badRequest", res);
+    return;
+  }
+
+  // Check if there is an existing match waiting
+  const matchExists = await Match.findOne({ difficulty: difficulty });
+  if (!matchExists) {
+    const match = new Match({
+      email: email,
+      difficulty: difficulty,
+      socketId: socket.id,
+    });
+
+    await match.save();
+    return;
+  }
```

Comment:
How are we ensuring that `Match.findOne()` is called repeatedly during the search, seems like it only tries to find once

Reply:
There is no need to ensure that it is called repeatedly
We just need to make sure that every person does a findOne. 

E.g
A put in an request to find a match with difficulty of easy 
A does a findOne and there is no existing request with difficulty easy
B put in an request to find a match with difficulty of hard
B does a findOne and no existing request with the same difficulty

B has already done a findOne , A and B is not matched. 
A does not need to trigger a findOne as only 1 more request is made which is B.
This case can be extend to multiple people.